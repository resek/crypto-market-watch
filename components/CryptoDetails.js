import { inject, observer } from 'mobx-react';
import { withRouter } from 'next/router';
import { Fragment, Component } from "react";

class CryptoDetails extends Component {

    render() {

        const id = this.props.router.query.id;
        const selectedCurrency = this.props.store.currency;
        const selectedSymbol = this.props.store.currencySymbol(selectedCurrency);
        const store = this.props.store;

        return (
            <Fragment>
                {store.apiData.map(currency => {
                    if (id == currency.id) {
                        return(
                            <div key={currency.id}>
                                
                                <div>
                                    <a href={`/currency/${currency.id}`}><button>Refresh data</button></a>
                                </div>                            
                                
                                <ul>
                                    <li><b>Rank:</b> {currency.rank}</li>
                                    <li><b>Name:</b> {currency.name}</li>
                                    <li><b>Symbol:</b> {currency.symbol}</li>
    
                                    <li><b>Price:</b> {selectedSymbol}{store.priceDecimal(eval(`currency.quote.price${selectedCurrency}`)).toLocaleString()}</li>
    
                                    <li><b>Volume 24h:</b> {selectedSymbol}{(eval(`currency.quote.volume24h${selectedCurrency}`)).toLocaleString().split('.')[0]}</li>
    
                                    <li><b>Market Cap:</b> {selectedSymbol}{(eval(`currency.quote.marketCap${selectedCurrency}`)).toLocaleString().split('.')[0]}</li>

                                    <li><b>Price in Bitcoin:</b> {(currency.id == 1) ? "1.00000000" : currency.quote.priceBTC.toFixed(8) } BTC</li>

                                    <li><b>Change 1h:</b> {currency.quote.change1h.toFixed(2)}%</li>
                                    <li><b>Change 24h:</b> {currency.quote.change24h.toFixed(2)}%</li>
                                    <li><b>Change 7d:</b> {currency.quote.change7d.toFixed(2)}%</li>

                                    <li><b>Total supply:</b> {currency.totalSupply.toLocaleString().split('.')[0]} {currency.symbol}</li>

                                    <li><b>Circulating supply:</b> {currency.circulatingSupply.toLocaleString().split('.')[0]} {currency.symbol}</li>
                                </ul>
                            </div>
                        )
                    }
                })}
                <style jsx>{`
                    div div {
                        width: 300px;
                        margin: 0 auto;
                        text-align: center;
                    }
                    ul {
                        width: 300px;
                        margin: 20px auto 0 auto;
                        list-style-type: none;
                        padding-left: 0;
                    }
                    li {
                        margin-bottom: 10px;
                    }
                `}</style>
            </Fragment>
        )
    }   
};

export default inject("store")(withRouter(observer(CryptoDetails)));