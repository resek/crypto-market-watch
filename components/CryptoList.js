import { Component } from "react";
import { inject } from 'mobx-react';
import Link from 'next/link';

class CryptoList extends Component {
    
    render () {
    
        return (
            <table>
                <thead>
                    <tr>
                        <th>Rank</th>
                        <th>Symbol</th>
                        <th>Price</th>
                        <th>Volume 24h</th>
                    </tr>
                </thead>
                <tbody>
                    {this.props.store.apiData.map((currency) => (
                        <tr key={currency.id}>
                            <td>{currency.rank}</td>
                            <td><Link as={`/currency/${currency.id}`} href={`/currency?id=${currency.id}`}><a>{currency.symbol}</a></Link></td>
                            <td>{currency.quote.priceUSD.toFixed(2)}$</td>
                            <td>{currency.quote.change24h.toFixed(2)}%</td>
                        </tr>
                    ))}
                </tbody>
                <style jsx>{`
                    table {
                        margin: 0 auto;
                    }
                    td {
                        width: 80px;
                        text-align: center;
                        height: 30px;
                    }
                `}</style>    
            </table>
        )
    }    
}

export default inject("store")(CryptoList);