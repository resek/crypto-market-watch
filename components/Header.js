import Link from 'next/link'

const linkStyle = {
  marginRight: 15,
}

const headerStyle = {
    marginBottom: 20,
}

const Header = () => (
    <div style={headerStyle}>
        <Link href="/">
          <a style={linkStyle}>Home</a>
        </Link>
        <Link href="#">
          <a style={linkStyle}>Refresh</a>
        </Link>
    </div>
)

export default Header