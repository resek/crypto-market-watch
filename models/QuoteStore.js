import {types} from "mobx-state-tree";

const QuoteStore = types
    .model("QuoteStore", {  
        change1h: types.number,
        change24h: types.number,
        change7d: types.number,
        priceEUR: types.number,        
        priceUSD: types.number, 
        priceCNY: types.number,
        priceBTC: types.number,
        volume24hEUR: types.number,
        volume24hUSD: types.number,
        volume24hCNY: types.number,
        marketCapEUR: types.number,
        marketCapUSD: types.number,
        marketCapCNY: types.number,
    });

export default QuoteStore;