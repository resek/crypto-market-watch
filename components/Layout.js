import Header from './Header'

const layoutStyle = {
	margin: 20,
}

const Layout = (props) => (
  	<div style={layoutStyle}>
      	<Header />
      	{props.children}
    </div>
)

export default Layout