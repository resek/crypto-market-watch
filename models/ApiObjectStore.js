import {types} from "mobx-state-tree";
import QuoteStore from "./QuoteStore";

const ApiObjectStore = types
    .model("ApiObjectStore", {
        id: types.number,
        name: types.string,
        symbol: types.string,
        rank: types.number,
        totalSupply: types.number,
        circulatingSupply: types.number,
        quote: types.frozen(QuoteStore),
    });

export default ApiObjectStore;