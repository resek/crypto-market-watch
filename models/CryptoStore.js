import { types, applySnapshot } from 'mobx-state-tree';
import axios from "axios";
import getConfig from 'next/config';
import ApiObjectStore from "./ApiObjectStore";

const {serverRuntimeConfig} = getConfig()

let store = null;

const CryptoStore = types
    .model("CryptoStore", {
    	apiData: types.array(ApiObjectStore)
    })
    .actions(self => ({
        async getApiData() {
            const response = await axios({
                method: 'GET',
                url: "https://pro-api.coinmarketcap.com/v1/cryptocurrency/listings/latest",
                headers: {
                    'X-CMC_PRO_API_KEY': serverRuntimeConfig.apiKey,
                },
                json: true,
                gzip: true
            });
            self.saveData(response.data.data);
        },
        saveData(response) {           
            response.map(currency => {
                const object = {id: currency.id, name: currency.name};
                self.apiData.push(object);
            })
        }
    }));

export function initStore (snapshot = null) {
    if (store === null) {
        store = CryptoStore.create({ apiData: [] })
    } 
    if (snapshot) {        
        applySnapshot(store, snapshot)
    }
    return store
}