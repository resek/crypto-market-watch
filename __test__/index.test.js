import { shallow } from 'enzyme';
import React from 'react';

import { initStore } from '../models/CryptoStore';
import Index from '../pages/index';
import { HeaderTest } from "../components/Header";
import { CryptoListTest } from "../components/CryptoList";

describe('index page snapshots', () => {    

    it("check index page snapshot", () => {
        const wrapper = shallow(<Index />);
        expect(wrapper).toMatchSnapshot();
    });

    it("check Header component snapshot", () => {
        const wrapper = shallow(<HeaderTest />);
        expect(wrapper).toMatchSnapshot();
    });

    it("check CryptoList component snapshot", () => {
        const store = initStore(true, null, false);
        const wrapper = shallow(<CryptoListTest store={store} />);
        expect(wrapper).toMatchSnapshot();
    });
});

describe("check UI", () => {
    
    it("Header has 3 buttons", () => {
        const wrapper = shallow(<HeaderTest />);
        expect(wrapper.find("button")).toHaveLength(3);
    });

    it("CryptoList contains refresh button", () => {
        const store = initStore(true, null, false);
        const wrapper = shallow(<CryptoListTest store={store} />);
        expect(wrapper.text().includes('Refresh data')).toBe(true);
    });
});