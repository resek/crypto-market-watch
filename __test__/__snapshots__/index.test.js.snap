// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index page snapshots check CryptoList component snapshot 1`] = `
<div
  className="jsx-3037929197"
>
  <a
    className="jsx-3037929197"
    href="/"
  >
    <button
      className="jsx-3037929197"
    >
      Refresh data
    </button>
  </a>
  <table
    className="jsx-3037929197"
  >
    <thead
      className="jsx-3037929197"
    >
      <tr
        className="jsx-3037929197"
      >
        <th
          className="jsx-3037929197"
        >
          Rank
        </th>
        <th
          className="jsx-3037929197"
        >
          Symbol
        </th>
        <th
          className="jsx-3037929197"
        >
          Price
        </th>
        <th
          className="jsx-3037929197"
        >
          Change 24h
        </th>
      </tr>
    </thead>
    <tbody
      className="jsx-3037929197"
    />
  </table>
  <JSXStyle
    css="div.jsx-3037929197{text-align:center;}table.jsx-3037929197{margin:20px auto 0 auto;}td.jsx-3037929197{width:90px;text-align:center;height:30px;}"
    styleId="3037929197"
  />
</div>
`;

exports[`index page snapshots check Header component snapshot 1`] = `
<div
  className="jsx-1868560903"
>
  <Link
    href="/"
  >
    <a
      className="jsx-1868560903"
    >
      Home
    </a>
  </Link>
  <button
    className="jsx-1868560903"
    onClick={[Function]}
  >
    USD
  </button>
  <button
    className="jsx-1868560903"
    onClick={[Function]}
  >
    EUR
  </button>
  <button
    className="jsx-1868560903"
    onClick={[Function]}
  >
    CNY
  </button>
  <JSXStyle
    css="div.jsx-1868560903{margin-bottom:25px;text-align:center;}button.jsx-1868560903{margin-left:10px;}"
    styleId="1868560903"
  />
</div>
`;

exports[`index page snapshots check index page snapshot 1`] = `
<Provider
  store={
    Object {
      "apiData": Array [],
      "currency": "USD",
      "exchRateBTC": 0,
      "exchRateCNY": 0,
      "exchRateUSD": 0,
    }
  }
>
  <Layout>
    <inject-CryptoListComp-with-store />
  </Layout>
</Provider>
`;
